drivers/posixaio-backend.c

```diff
@@ -0,0 +1,598 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the names of its 
+ *     contributors may be used to endorse or promote products derived from 
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
+ * OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#define IO_SIGNAL SIGUSR1
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <aio.h>
+
+#include "tapdisk.h"
+#include "tapdisk-log.h"
+#include "posixaio-backend.h"
+#include "tapdisk-server.h"
+#include "tapdisk-utils.h"
+#include "timeout-math.h"
+#include <signal.h>
+
+#include "atomicio.h"
+#include "aio_getevents.h"
+#include <sys/signalfd.h>
+#include "debug.h"
+
+
+#define WARN(_f, _a...) tlog_write(TLOG_WARN, _f, ##_a)
+#define DBG(_f, _a...) tlog_write(TLOG_DBG, _f, ##_a)
+#define ERR(_err, _f, _a...) tlog_error(_err, _f, ##_a)
+
+#define posixaio_backend_queue_empty(q) ((q)->queued == 0)
+#define posixaio_backend_queue_full(q)  \
+	(((q)->tiocbs_pending + (q)->queued) >= (q)->size)
+typedef struct _posix_aio_queue {
+	int                   size;
+
+	const struct tio     *tio;
+	void                 *tio_data;
+
+	int                   queued;
+	struct aiocb        **aiocbList;
+	struct tiocb       **tiocbList;
+
+	int                   tiocbs_pending;
+	struct tlist          pending;
+
+	struct tlist          deferred;
+	int                   tiocbs_deferred;
+
+	uint64_t              deferrals;
+} posix_aio_queue;
+
+struct tio {
+	const char           *name;
+	size_t                data_size;
+
+	int  (*tio_setup)    (posix_aio_queue *queue, int qlen);
+	void (*tio_destroy)  (posix_aio_queue *queue);
+	int  (*tio_submit)   (posix_aio_queue *queue);
+};
+
+enum {
+	TIO_DRV_LIO     = 1,
+};
+
+static inline void
+queue_tiocb(posix_aio_queue *queue, struct tiocb *tiocb)
+{
+	queue->aiocbList[queue->queued] = (struct aiocb*)tiocb->iocb;
```

Comment:
Might be useful to have a bounds check here, at least as an assert. One of the callers does a check whether the queue is full or not before calling or not, but the other one doesn't seem to.